drgn_tools/lsmod.py

```diff
@@ -0,0 +1,165 @@
+# Copyright (c) 2023, Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/
+import argparse
+
+from drgn import Object
+from drgn import Program
+from drgn.helpers.common.memory import identify_address
+from drgn.helpers.linux.list import list_for_each_entry
+
+from drgn_tools.corelens import CorelensModule
+
+
+def get_parm_valf(prog: Program, parm_val: str, kp: Object) -> int:
+    """
+    converting endian to big endian
+
+    :param prog: drgn program
+    :param parm_val: kernel_param.arg is the paramter value that is converted to big endian
+    :param kp: kernel_param
+    :returns: int after converting from hex
+    """
+    parm_val1 = ""
+    for i in range(0, len(parm_val), 2):
+        parm_val1 = (
+            str(prog.read(kp.arg.value_(), 2).hex())[i : i + 2] + parm_val1
+        )
+    return int(str(parm_val1), 16)
+
+
+def lsmod(prog: Program) -> None:
+    """
+    List loaded modules and their details.
+
+    :param prog: drgn program
+    :returns: None
+    """
+
+    # List all loaded modules
+    print(
+        "MODULE".ljust(25)
+        + "NAME".ljust(20)
+        + "\tSIZE".rjust(15)
+        + "\tREF".rjust(3)
+        + "\tDEPENDENDENT MODULES".rjust(15)
+    )
+    print("-" * 90)
+    for use in list_for_each_entry(
+        "struct module", prog["modules"].address_of_(), "list"
+    ):
+        dep_mod = []
+        for depuse in list_for_each_entry(
+            "struct module_use", use.source_list.address_of_(), "source_list"
+        ):
+            dep_mod.append(depuse.source.name.string_().decode("utf-8"))
+        print(
+            hex(use.mkobj.mod.value_()).ljust(5),
+            "\t",
+            use.name.string_().decode("utf-8").ljust(15),
+            "\t",
+            str(use.core_layout.size.value_()).rjust(10),
+            str(use.refcnt.counter.value_() - 1).rjust(6),
+            "\t",
+            ", ".join(dep_mod),
+        )
+
+    # Module details
+    for use in list_for_each_entry(
+        "struct module", prog["modules"].address_of_(), "list"
+    ):
+        print("\n")
+        print("MODULE NAME:".ljust(15), use.name.string_().decode("utf-8"))
+        print("PARAM COUNT:", str(use.num_kp.value_()))
+        print("ADDRESS    :", hex(use.num_kp.address_of_()))
+        print(
+            "PARAMETER".ljust(25),
+            "ADDRESS".rjust(15),
+            "TYPE".rjust(25),
+            "VALUE".rjust(10),
+        )
+        ker_parm = hex(use.kp.value_())
+        for i in range(0, use.num_kp.value_()):
+            parm_valf = ""
+            parm_val = ""
+            kp = Object(prog, "struct kernel_param", address=int(ker_parm, 16))
```

Comment:
It looks like you're having a little trouble iterating over this array. You can do it like this, and not need to maintain your manual ker_parm variable:
```suggestion
        for i in range(use.num_kp):
            parm_valf = ""
            parm_val = ""
            kp = use.kp[i]
```